### ISSUE TITLE
Rewrite Worker-Scheduler relationship

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#181]
### Problem

Many small tasks aren't running as quickly as possible.  E.g.

```python
In [1]: from distributed import Executor, wait
In [2]: e = Executor('192.168.1.141:8786')  # full cluster is on local machine

In [3]: def inc(x):
    return x + 1

In [4]: %time _ = wait(e.map(inc, range(10000)))
CPU times: user 3.19 s, sys: 216 ms, total: 3.4 s
Wall time: 6.5 s

In [5]: %time _ = [inc(i) for i in range(10000)]
CPU times: user 124 ms, sys: 7.72 ms, total: 132 ms
Wall time: 129 ms
```

This is the active bottleneck for fast computations like `df.column.sum()`.  Generally we're limited to 1ms per task on this, which can easily grow into the seconds for moderately sized datasets.

I think that this can be resolved by keeping more tasks in flight and by batching more tasks together.

### Current state

Currently we keep an active connection for every core on every worker node.  As the core becomes free the scheduler finds a new task, sends it across the wire, and then waits on the acknowledgment of completion before repeating.

There are a few things wrong with this.  The worker CPU idles during network communication and while the scheduler finds an appropriate task.  If latency time is much greater than computation time, then we risk spending most of our time waiting on network traffic.  There are also secondary concerns of multiple file descriptors and serialization costs.

### Proposal

So we should probably have an asynchronous pipeline from the scheduler to the worker that sends tasks and receives tasks in a less lock-step fashion.  Additionally, we should probably consider sending multiple tasks in a single message and similarly consider sending multiple acknowledgements in a single message.

On the acknowledgement side, e.g.  "Hi scheduler, I, a worker, just finished these three tasks" I think we can just institute a policy of "don't send more than one communication every n milliseconds" where `n` is something like 1-10.  There is a tradeoff here between increasing latency and batching to increase bandwidth.  I think it's fine if we add a few ms in the case where we're sending off a large number of acknowledgments.

On the submission side I'm still unsure of the right approach.  At any given time we can pull from two sources of tasks

1.  The stack of tasks that would prefer to be run on our worker, likely due to data locality
2.  The queue of tasks shared by all workers

We want to send enough tasks so that the worker CPUs remain saturated, i.e. so that there is enough of a backlog of tasks on the worker so that it doesn't run dry during network communication.  However, we don't want to send too many tasks too early to the worker, because once they're sent we can't easily decide to computing something better instead.  The tradeoff here is between saturating the CPU and optimal task selection.  Additionally, we don't want to pull too much from the common queue, which is generally shared by all workers.

Finally, there is the concern of wildly differing task duration.  We might run many `inc` computations, learn that we need to keep 100 tasks on the worker to keep it occupied, and then accidentally send 100 very long running computations all at once.

I'll be thinking about this for a bit, thought others might be interested.

**Comment by mrocklin** [comment]
To be more explicit, the problem outlined towards the end is, I think, a control problem.  

There is a flow of tasks into and out of the worker.  We want there to be as many tasks in the worker as there are cores (e.g. eight).  Being low is very bad (under-utilization), being high is a little bad (inefficient scheduling).  We can control the flow of tasks in, but with some lag.  The worker itself will probably churn through tasks at piecewise constant speeds (we tend to receive many tasks of the same type, then many of another.)

What are good systems from which to gain insight.  PID controllers?

cc @jcrist @eriknw @seibert

**Comment by eriknw** [comment]
Good systems, eh?  CDC 6600? ;)

**Comment by mrocklin** [comment]
OK, so I think the following might work:  

We have a basic PI controller on the scheduler to ramps up flow of tasks to the worker when we're in the many-small tasks situation (the only case when simple synchronous scheduling doesn't work.).  

This works great until we switch from very fast tasks to very slow tasks, when we risk dumping 1000s of long running tasks onto a single worker.  This risk is directly proportional to the ratio of the scheduler-worker latency over the task length.  For example if roundtrip times are 10ms and tasks take 10us and we want 10 tasks on the machine at all times, then we will want to keep something like 10000 tasks in the pipe flowing to the worker.  

When we switch from fast tasks to slow tasks we risk saturating the worker with many long-running tasks.  Fortunately, I think we can just send them back.  Return flow should, I think, solve this problem at a maximum cost of an extra roundtrip overhead on the extra tasks.
